X86/X86AdditionalInstrInfo.h

```diff
@@ -61,6 +62,9 @@ struct X86AdditionalInstrInfo {
   uint8_t MemOpSize;
   // Instruction kind
   InstructionKind InstKind;
+  // specifies the precision of operations that can take multiple precisions
+  // in same-width registers (e.g.)
+  uint8_t BitPrecision = 0;
```

Comment:
Thanks, I agree that adding another field was not the best solution and I'd like to remove it!

I could not find a method to get the bit precision from the operands. There is `getPhysRegSizeInBits(Op.getReg())`, but that returns `64` for both `float` and `double` instruction, as there are just 64-bit `XMM` registers. Is there another function that I've missed? Thanks :)

Reply:
Actually, an XMM register has 128 bits. That's why it can be used for both floats and doubles. I think that the decision of whether it represents a float or a double cannot be inferred by the register alone. XMM registers are different from those where we use different names of essentially the same register to denote its bit-length, e.g., RAX, EAX, AX, etc. The actual bit-length is defined per instruction, even if it is an RR instruction. For example, both the 'movaps' and  'movapd' are RR instructions that use XMM registers.
https://www.felixcloutier.com/x86/movaps
https://www.felixcloutier.com/x86/movapd
However, one of them should be used with floats and the other with doubles. When we represent 'movaps' simply using the InstructionKind::SSE_MOV_RR we cannot know the bit-length it operates on. 

That's why I think we need to specify both the InstructionKind and the bit-length that the instruction operates on.

Reply:
BTW, the same issue will arise when using XMM register for SIMD instructions. It can represent a SIMD vector of 16 'chars' of 8-bits each up to 4 lanes of 32-bit integers. It will only depend on which instruction we are using. For example, 'paddb xmm0, xmm1' is the addition that assumes we are operating on vectors with 16 lanes of 8-bits int, but 'paddd' assumes we are using vectors with 4 lanes of 32-bit ints.
https://mudongliang.github.io/x86/html/file_module_x86_id_226.html

I believe the same solution of specifying the bit-length of the operation will be necessary for SIMD instructions as well and we can reuse the same field.

Reply:
I agree with @bharadwajy that querying this information from the LLVM infrastructure will be great. Maybe mctoll could reconstruct this information based on this MachineValueType:
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/MachineValueType.h#L30

Reply:
Since the precision information is indicated by the instruction opcode (an not the operands), we may be able to peek into the [InstructionFormat](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86InstrFormats.td) to discern the instruction kind and precision information from [BaseInfo](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h). For example, note that X86II::XS and X86II::XD are used to indicate "single and double precision scalar floating point operations performed in the SSE registers". I am experimenting to see if that suffices.

Reply:
Thanks for that suggestion @bharadwajy! That seems to work.

I'll squash all of these commits into a single one once I've fixed all issues.

Reply:
Attached is the diff with an alternative way to get precision without adding the new field `uint8_t BitPrecision` to  `struct X86AdditionalinstrInfo`. The implementation looks at `TSFlags` of `MCInstrDesc` of a `MachineInstr`. A bit of clean up - such as getting rid of additions to the bit precision values added in this PR to the static table and the field declaration itself - is needed. 

Apply this diff to the tip of your fork's master branch.

All tests (including the new ones added in this PR) pass with these changes.

Let me know if you have any questions. I'll look at the rest of the changes in this PR soon.

Thanks!

[get-instr-precision.txt](https://github.com/microsoft/llvm-mctoll/files/6444472/get-instr-precision.txt)

Reply:
I applied your diff, seems a fair bit cleaner than the solution I came up with. Thanks! :)